src/hotspot/share/utilities/growableArray.hpp

```diff
@@ -481,40 +482,80 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
   void swap(GrowableArrayWithAllocator<E, Derived>* other) {
     ::swap(this->_data, other->_data);
     ::swap(this->_len, other->_len);
-    ::swap(this->_max, other->_max);
+    ::swap(this->_capacity, other->_capacity);
   }
 
+  // Ensure capacity is at least new_capacity.
+  void reserve(int new_capacity);
+
+  // Reduce capacity to length.
+  void shrink_to_fit();
+
   void clear_and_deallocate();
 };
 
 template <typename E, typename Derived>
-void GrowableArrayWithAllocator<E, Derived>::grow(int j) {
-  int old_max = this->_max;
-  // grow the array by increasing _max to the first power of two larger than the size we need
-  this->_max = next_power_of_2((uint32_t)j);
-  // j < _max
+void GrowableArrayWithAllocator<E, Derived>::expand_to(int new_capacity) {
+  int old_capacity = this->_capacity;
+  assert(new_capacity > old_capacity,
+         "expected growth but %d <= %d", new_capacity, old_capacity);
+  this->_capacity = new_capacity;
   E* newData = static_cast<Derived*>(this)->allocate();
   int i = 0;
   for (     ; i < this->_len; i++) ::new ((void*)&newData[i]) E(this->_data[i]);
-  for (     ; i < this->_max; i++) ::new ((void*)&newData[i]) E();
-  for (i = 0; i < old_max; i++) this->_data[i].~E();
+  for (     ; i < this->_capacity; i++) ::new ((void*)&newData[i]) E();
+  for (i = 0; i < old_capacity; i++) this->_data[i].~E();
   if (this->_data != NULL) {
     static_cast<Derived*>(this)->deallocate(this->_data);
   }
   this->_data = newData;
 }
 
 template <typename E, typename Derived>
-void GrowableArrayWithAllocator<E, Derived>::clear_and_deallocate() {
-  if (this->_data != NULL) {
-    for (int i = 0; i < this->_max; i++) {
-      this->_data[i].~E();
-    }
-    static_cast<Derived*>(this)->deallocate(this->_data);
-    this->_data = NULL;
+void GrowableArrayWithAllocator<E, Derived>::grow(int j) {
+  // grow the array by increasing _capacity to the first power of two larger than the size we need
+  expand_to(next_power_of_2((uint32_t)j));
+}
+
+template <typename E, typename Derived>
+void GrowableArrayWithAllocator<E, Derived>::reserve(int new_capacity) {
+  if (new_capacity > this->_capacity) {
+    expand_to(new_capacity);
   }
-  this->_len = 0;
-  this->_max = 0;
+}
+
+template <typename E, typename Derived>
+void GrowableArrayWithAllocator<E, Derived>::shrink_to_fit() {
+  int old_capacity = this->_capacity;
+  int len = this->_len;
+  assert(len <= old_capacity, "invariant");
+
+  // If already at full capacity, nothing to do.
+  if (len == old_capacity) {
+    return;
+  }
+
+  // If not empty, allocate new, smaller, data, and copy old data to it.
+  E* old_data = this->_data;
+  E* new_data = nullptr;
+  this->_capacity = len;        // Must preceed allocate().
+  if (len > 0) {
+    new_data = static_cast<Derived*>(this)->allocate();
+    for (int i = 0; i < len; ++i) ::new (&new_data[i]) E(old_data[i]);
```

Comment:
This can be a stupid question as I'm confused a little bit.
Why do we reallocate memory for data elements?
Could we just move the element pointers from the `old_data`?